sklearn/utils/_array_api.py

```diff
@@ -261,19 +263,51 @@ def _isdtype_single(dtype, kind, *, xp):
         return dtype == kind
 
 
-def supported_float_dtypes(xp):
-    """Supported floating point types for the namespace.
+def supported_float_dtypes(xp, device=None):
+    """Supported floating point types for the namespace/device pair.
 
-    Note: float16 is not officially part of the Array API spec at the
+    Parameters
+    ----------
+    xp : module
+        Array namespace to inspect.
+
+    device : str, default=None
+        Device to use for dtype selection. If ``None``, then a default device
+        is assumed.
+
+    Returns
+    -------
+    supported_dtypes : tuple
+        Tuple of real floating data types supported by the provided array namespace,
+        ordered from the highest precision to lowest.
+
+    See Also
+    --------
+    max_precision_float_dtype : Maximum float dtype for a namespace/device pair.
+
+    Notes
+    -----
+    `float16` is not officially part of the Array API spec at the
     time of writing but scikit-learn estimators and functions can choose
     to accept it when xp.float16 is defined.
 
+    Additionally, some devices available within a namespace may not support
+    all floating-point types that the namespace provides.
+
     https://data-apis.org/array-api/latest/API_specification/data_types.html
     """
-    if hasattr(xp, "float16"):
-        return (xp.float64, xp.float32, xp.float16)
+    # TODO: Update to use `__array_namespace__info__()` from array-api v2023.12
+    #       when/if that becomes more widespread.
+    if xp.__name__ in {"array_api_compat.torch", "torch"} and device == "mps":
+        # N.B. Yanked from pull/27232
```

Comment:
I am not sure if we still want to make `supported_float_dtypes` device sensitive: the new `_max_precision_float_dtype` in `main` does not delegate to `supported_float_dtypes` so we do not really have a case for this change anymore.